ml-agents/mlagents/trainers/ppo/trainer.py

```diff
@@ -451,19 +452,26 @@ def update_policy(self):
             ):
                 start = l * n_sequences
                 end = (l + 1) * n_sequences
-                run_out = self.policy.update(
-                    buffer.make_mini_batch(start, end), n_sequences
-                )
-                value_total.append(run_out["value_loss"])
-                policy_total.append(np.abs(run_out["policy_loss"]))
-        self.stats["Losses/Value Loss"].append(np.mean(value_total))
-        self.stats["Losses/Policy Loss"].append(np.mean(policy_total))
-        for _, reward_signal in self.policy.reward_signals.items():
-            update_stats = reward_signal.update(
-                self.training_buffer.update_buffer, n_sequences
-            )
-            for stat, val in update_stats.items():
-                self.stats[stat].append(val)
+                mini_batch = buffer.make_mini_batch(start, end)
+
+                feed_dict = {}
+                update_dict = self.policy.update_dict
+                feed_dict.update(self.policy.prepare_update(mini_batch, n_sequences))
+                stats_needed = self.policy.stats_name_to_update_name
+                for _, reward_signal in self.policy.reward_signals.items():
```

Comment:
Will it be better moving this part (constructing feed_dict and update_dict) into policy, since all the reward_signals and stats_name_to_update_name are stored in policy? Maybe keep the policy update function and move it there?

Reply:
Good idea, I think bringing back the update() method makes sense. This would also make it so that we don't have to change the BC trainer to keep the Policy interface the same, since the BC trainer has no reward signals.